src/SpotlessDiagnostics.ts

```diff
@@ -0,0 +1,206 @@
+import * as vscode from 'vscode';
+import {
+  Difference,
+  generateDifferences,
+  showInvisibles,
+} from 'prettier-linter-helpers';
+import { Spotless } from './Spotless';
+import { logger } from './logger';
+import { debounce } from './util';
+import { SpotlessRunner } from './SpotlessRunner';
+
+const DIAGNOSTICS_UPDATES_DEBOUNCE_MS = 0;
+
+export interface SpotlessDiff {
+  source: string;
+  formattedSource: string;
+  differences: Difference[];
+}
+
+export class SpotlessDiagnostics {
+  private diagnosticCollection: vscode.DiagnosticCollection;
+  private getDiffPromise: Promise<SpotlessDiff> | undefined;
+  private currentDiff: SpotlessDiff | undefined;
+  private diagnosticMap: Map<string, vscode.Diagnostic[]> = new Map();
+  private isStale = false;
+
+  constructor(
+    private readonly context: vscode.ExtensionContext,
+    private readonly spotless: Spotless,
+    private readonly spotlessRunner: SpotlessRunner
+  ) {
+    this.diagnosticCollection = vscode.languages.createDiagnosticCollection(
+      'java'
+    );
+  }
+
+  public register(): void {
+    const onReady = this.spotless.onReady(() => {
+      const activeDocument = vscode.window.activeTextEditor?.document;
+      if (activeDocument) {
+        void this.runDiagnostics(activeDocument);
+      }
+    });
+
+    const onDidChangeTextDocument = vscode.workspace.onDidChangeTextDocument(
+      (e: vscode.TextDocumentChangeEvent) => {
+        if (
+          e.contentChanges.length &&
+          vscode.window.activeTextEditor?.document === e.document
+        ) {
+          void this.handleChangeTextDocument(e.document);
+        }
+      }
+    );
+
+    const onDidChangeActiveTextEditor = vscode.window.onDidChangeActiveTextEditor(
+      (editor?: vscode.TextEditor) => {
+        if (editor) {
+          void this.runDiagnostics(editor.document);
+        }
+      }
+    );
+
+    this.context.subscriptions.push(
+      onReady,
+      onDidChangeTextDocument,
+      onDidChangeActiveTextEditor,
+      this.diagnosticCollection
+    );
+  }
+
+  @debounce(DIAGNOSTICS_UPDATES_DEBOUNCE_MS)
+  async handleChangeTextDocument(document: vscode.TextDocument): Promise<void> {
+    // TODO: prevent diagnostics if run after applying formatting
+    void this.runDiagnostics(document);
+  }
+
+  public async runDiagnostics(
+    document: vscode.TextDocument,
+    cancellationToken?: vscode.CancellationToken
+  ): Promise<SpotlessDiff | undefined> {
+    if (document.languageId.toLowerCase() !== this.diagnosticCollection.name) {
+      return;
+    }
+    // There's already a diagnostic session running, mark the current one as stale
+    if (this.getDiffPromise) {
+      this.isStale = true;
```

Comment:
can we make this stale logic generic and build it into the SpotlessRunner to benefit both diagnostics and formatting